dao/tenant_dao.go

```diff
@@ -103,3 +111,141 @@ func (t *tenantDaoImpl) TenantByIdentity(id *identity.Identity) (*m.Tenant, erro
 
 	return &tenant, nil
 }
+
+func (t *tenantDaoImpl) GetUntranslatedTenants() ([]m.Tenant, error) {
+	var tenants []m.Tenant
+
+	err := DB.Debug().
+		Model(&m.Tenant{}).
+		Where(untranslatedTenantsWhereCondition).
+		Find(&tenants).
+		Error
+
+	if err != nil {
+		return nil, err
+	}
+
+	return tenants, err
+}
+
+func (t *tenantDaoImpl) TranslateTenants() (int64, uint64, uint64, []m.TenantTranslation, error) {
+	// Count the total number of translatable tenants.
+	var translatableTenants int64
+	// Count the translation operations.
+	var translatedTenants, untranslatedTenants uint64
+	// Keep track of each of the tenant translations.
+	var tenantTranslations []m.TenantTranslation
+
+	// Run everything inside a transaction to avoid inconsistencies. This way if new tenants are added/removed while
+	// the translation is in progress, all the counts and the results will still match.
+	err := DB.Debug().Transaction(func(tx *gorm.DB) error {
+		// Get the count to be able to process the tenants in batches of 100 tenants.
+		err := tx.
+			Model(&m.Tenant{}).
+			Where(untranslatedTenantsWhereCondition).
+			Count(&translatableTenants).
+			Error
+
+		if err != nil {
+			return fmt.Errorf("unable to count untranslated tenants: %w", err)
+		}
+
+		translator := tenantid.NewTranslator(config.Get().TenantTranslatorUrl)
+		for i := int64(0); i < translatableTenants; i += 100 {
+			// Grab all the EBS account numbers to be processed.
+			var ebsAccountNumbers []string
+			err := tx.
+				Model(&m.Tenant{}).
+				Where(untranslatedTenantsWhereCondition).
+				Pluck("external_tenant", &ebsAccountNumbers).
+				Offset(int(i)).
+				Order(clause.OrderByColumn{Column: clause.Column{Name: "external_tenant"}, Desc: false}).
+				Limit(100).
+				Error
+
+			if err != nil {
+				return fmt.Errorf("unable to fetch the tenants to translate them: %w", err)
+			}
+
+			// Batch process 100 EBS account numbers.
+			results, err := translator.EANsToOrgIDs(context.Background(), ebsAccountNumbers)
```

Comment:
This will actually send all of the `ebsAccountNumbers` at once, the name of the method is misleading.

We should probably loop over this slice 100 at a time and translate them that way.

Reply:
I believe I am doing that in the code, right? Everything is wrapped on a `for i := int64(0); i < translatableTenants; i += 100` loop. Were you actually referring to fetching everything at once and then applying the loop to just the translator part?

Reply:
Yep I see it now. I totally missed it on L154, sorry about that.

Reply:
No problem at all! :)